### DISCUSSION TITLE
How here to extends componets in the right way

### DISCUSSION THREAD

**Discussion by andrey-hilux** [quasarframework/quasar#9735] [General - Components / Directives / etc]
Typically you have something like:
```js
setup (props, { emit, attrs }) {  
  ...
   return h('label', {...}) 
}
```
Is there a way to extends through the setup function?

**Comment by ontwikkelfabriek** [comment]
@andrey-hilux 

> How here to extends componets in the right way

I think they made extending components extra hard in vue3 compared to vue2. ( `extends`  not working as it used to )

Here's a discussion about it with some working solutions:
https://github.com/quasarframework/quasar/discussions/8761#discussioncomment-786781

I also found this thread that suggests an `init` function:
https://stackoverflow.com/questions/65020198/extend-component-with-composition-api-in-vue-js

**Reply by andrey-hilux** [reply]
No, [native solution](https://andrey-hilux.github.io/video/docs/bugs-upgrade-guide-quasar-v-2-to-v-3.html#4-extends-option):

```js
export default {
    extends: QInput,
    setup(props, setupContext) {
        const renderFn = QInput.setup(props, setupContext)
        // expose public methods
        const vm = getCurrentInstance()
        Object.assign(vm.proxy, {
            // focus,
            // select,
            // getNativeElement: () => inputRef.value
        })
        return renderFn
    },
    // ...
}
```

And I wish to know what to do with props

The legacy was with $options

**Reply by ontwikkelfabriek** [reply]
@andrey-hilux 

Ah very nice. But what about the **Options API** that 95% of the Vue developers are still using?

**Reply by andrey-hilux** [reply]
Options API? It was actual for VueJs 2 developers! 
Are you assuming the `QInput.options`?
In VueJs 3 this mechanic has become more easy and elegant. But with a bench of specific undocumented things and behaviors, which you can see only when it's debugging.

**Reply by ontwikkelfabriek** [reply]
> Options API? It was actual for VueJs 2 developers!

Ah I thought it was about Vue3. In Vue2 you could simply use `extends` with the option API to extend a component. Unfortunately `extends` does nothing anymore in Vue3 / Qv2...

**Reply by andrey-hilux** [reply]
Why? It works, but unfortunately, the quasar is not so friendly for things like this which encapsulated with only a few public functions and fields useless for me...

**Reply by andrey-hilux** [reply]
But generally, it's not a problem, if you can't do this, do it more easily!
